test/e2e/helpers.go

```diff
@@ -757,34 +754,47 @@ func getImageSkusInOffer(ctx context.Context, location, publisher, offer string)
 	imagesClient := compute.NewVirtualMachineImagesClient(subscriptionID)
 	imagesClient.Authorizer = authorizer
 
-	Byf("Finding image skus for offer %s/%s in %s", publisher, offer, location)
+	return imagesClient
+}
 
-	res, err := imagesClient.ListSkus(ctx, location, publisher, offer)
+// getVersionsInOffer returns a map of Kubernetes versions as strings to semver.Versions.
+func getVersionsInOffer(ctx context.Context, location, publisher, offer string) map[string]semver.Version {
+	Byf("Finding image skus and versions for offer %s/%s in %s", publisher, offer, location)
+	imagesClient := newImagesClient()
+	skusList, err := imagesClient.ListSkus(ctx, location, publisher, offer)
 	Expect(err).NotTo(HaveOccurred())
 
-	var skus []string
-	if res.Value != nil {
-		skus = make([]string, len(*res.Value))
-		for i, sku := range *res.Value {
-			// we have to do this to make sure the SKU has existing images
-			// see https://github.com/Azure/azure-cli/issues/20115.
+	var skus map[string]*compute.ListVirtualMachineImageResource
+	if skusList.Value != nil {
+		skus = make(map[string]*compute.ListVirtualMachineImageResource, len(*skusList.Value))
+		for _, sku := range *skusList.Value {
 			res, err := imagesClient.List(ctx, location, publisher, offer, *sku.Name, "", nil, "")
 			Expect(err).NotTo(HaveOccurred())
+			// Don't use SKUs without existing images. See https://github.com/Azure/azure-cli/issues/20115.
 			if res.Value != nil && len(*res.Value) > 0 {
-				skus[i] = *sku.Name
+				skus[*sku.Name] = &res
 			}
 		}
 	}
-	return skus
-}
 
-// parseImageSkuNames parses SKU names in format "k8s-1dot17dot2-os-123" to extract the Kubernetes version.
-// it returns a sorted list of all k8s versions found.
-func parseImageSkuNames(skus []string) map[string]semver.Version {
-	capiSku := regexp.MustCompile(`^k8s-(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)-[a-z]*.*$`)
+	capiSku := regexp.MustCompile(`^[\w-]+-gen[12]$`)
+	capiVersion := regexp.MustCompile(`^(\d)(\d{1,2})\.(\d{1,2})\.\d{8}$`)
+	oldCapiSku := regexp.MustCompile(`^k8s-(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)-[a-z]*.*$`)
 	versions := make(map[string]semver.Version, len(skus))
-	for _, sku := range skus {
+	for sku, res := range skus {
```

Comment:
nit: instead of storing the SKUs in a new map and iterating over the list of SKUs twice, you could move the processing directly into the block above so we build the list of versions as we go through the list of SKUs and image versions returned by the client.